lib/remote/objectqueryhandler.cpp

```diff
@@ -257,6 +261,63 @@ bool ObjectQueryHandler::HandleRequest(
 			if (!joinedObj)
 				continue;
 
+			Type::Ptr reflectionType = joinedObj->GetReflectionType();
+			Expression *permissionFilter = nullptr;
+			Expression::Ptr evaluatedFilter;
+
+			auto it = evaluatedTypes.find(reflectionType.get());
+
+			if (it == evaluatedTypes.end()) {
+				String permission = "objects/query/" + reflectionType->GetName();
+
+				if (!FilterUtility::HasPermission(user, permission, &permissionFilter)) {
+					// The API user isn't authorized to access this relation, so just ignore it because there
+					// is no reason to raise an exception here since the actual object requested by the client
+					// has been found!!
+					evaluatedFilter = permissionFilter;
+
+					evaluatedTypes.insert({reflectionType.get(), std::make_pair(evaluatedFilter, false)});
+
+					continue;
+				}
+
+				evaluatedFilter = permissionFilter;
+				evaluatedTypes.insert({reflectionType.get(), std::make_pair(evaluatedFilter, true)});
+			} else {
+				bool granted;
+
+				std::tie(evaluatedFilter, granted) = it->second;
+
+				if (! granted) {
+					// Not authorized
+					continue;
+				}
+			}
+
+			auto relation = joinedRelations.find(joinedObj.get());
+
+			if (relation == joinedRelations.end()) {
+				ScriptFrame permissionFrame(false, new Namespace());
+
+				try {
+					if (!FilterUtility::EvaluateFilter(permissionFrame, evaluatedFilter.get(), joinedObj)) {
+						joinedRelations.insert({joinedObj.get(), false});
+
+						continue;
+					}
+				} catch (const ScriptError& err) {
+					joinedRelations.insert({joinedObj.get(), false});
+
+					// Nothing to do just continue
+					continue;
+				}
+
+				joinedRelations.insert({joinedObj.get(), true});
+			} else if (!relation->second) {
+				// Access denied
+				continue;
+			}
```

Comment:
Could be transformed in a similar way, but will stay a bit longer due to the exception handler.